File: android/src/main/kotlin/dev/masel/synced_bloc/NativeSyncSlave.kt
package dev.masel.synced_bloc

class NativeSyncSlave private constructor(val masterId: String) {
    companion object {
        private val nativeSyncSlaves: MutableList<NativeSyncSlave> = mutableListOf()

        fun withMasterId(masterId: String): NativeSyncSlave {
            val existing = nativeSyncSlaves.firstOrNull { it.masterId == masterId }
            return if (existing != null) {
                existing.sync() //So always trigger change listeners
                existing
            } else {
                val new = NativeSyncSlave(masterId)
                SyncedBlocPlugin.getMasterWithId(masterId)?.let(new::setup)
                nativeSyncSlaves.add(new)
                new
            }
        }

        internal fun onRegisterMaster(master: SyncMaster) {
            nativeSyncSlaves.firstOrNull { it.masterId == master.masterId }?.setup(master)
        }

        internal fun onUnregisterMaster(masterId: String) {
            nativeSyncSlaves.firstOrNull { it.masterId == masterId }?.syncMaster = null
        }

        internal fun onMasterChange(masterId: String) {
            nativeSyncSlaves.firstOrNull { it.masterId == masterId }?.sync()
        }
    }

    fun interface ChangeListener {
        fun onSyncedBlocStateChange()
    }

    private val changeListeners: MutableList<ChangeListener> = mutableListOf()
    fun addChangeListener(listener: ChangeListener) = changeListeners.add(listener)
    fun removeChangeListener(listener: ChangeListener) = changeListeners.remove(listener)

    // Transitions from null after short init delay
    var jsonState: String? = null
        private set

    private var syncMaster: SyncMaster? = null

    private fun setup(syncMaster: SyncMaster) {
        this.syncMaster = syncMaster
        sync()
    }

    private fun sync() {
        syncMaster?.getState(
            onSuccess = { res ->
                jsonState = (res as String)
                changeListeners.forEach { it.onSyncedBlocStateChange() }
            },
            onError = { throw Error() }
        )
    }

    fun addEvent(event: String) {
        syncMaster?.addEvent(event)
    }
}

File: ios/Classes/NativeSyncSlave.swift
public class NativeSyncSlave {
    class ChangeListener {
        let onSyncedBlocStateChange: () -> Void

        init(onSyncedBlocStateChange: @escaping () -> Void) {
            self.onSyncedBlocStateChange = onSyncedBlocStateChange
        }
    }

    let masterId: String
    private static var nativeSyncSlaves: [NativeSyncSlave] = []
    private var changeListeners: [ChangeListener] = []
    private var jsonState: String? = nil
    private var syncMaster: SyncMaster? = nil

    private init(masterId: String) {
        self.masterId = masterId
    }

    static func withMasterId(masterId: String) -> NativeSyncSlave {
        if let existing = nativeSyncSlaves.first(where: { $0.masterId == masterId }) {
            existing.sync() // So always trigger change listeners
            return existing
        } else {
            let new = NativeSyncSlave(masterId: masterId)
            if let master = SwiftSyncedBlocPlugin.getMasterWithId(masterId: masterId) {
                new.setup(syncMaster: master)
            }
            nativeSyncSlaves.append(new)
            return new
        }
    }

    internal static func onRegisterMaster(master: SyncMaster) {
        if let slave = nativeSyncSlaves.first(where: { $0.masterId == master.masterId }) {
            slave.setup(syncMaster: master)
        }
    }

    internal static func onUnregisterMaster(masterId: String) {
        if let slave = nativeSyncSlaves.first(where: { $0.masterId == masterId }) {
            slave.syncMaster = nil
        }
    }

    internal static func onMasterChange(masterId: String) {
        if let slave = nativeSyncSlaves.first(where: { $0.masterId == masterId }) {
            slave.sync()
        }
    }

    func addChangeListener(listener: ChangeListener) {
        changeListeners.append(listener)
    }

    func removeChangeListener(listener: ChangeListener) {
        if let index = changeListeners.firstIndex(where: { $0 === listener }) {
            changeListeners.remove(at: index)
        }
    }

    private func setup(syncMaster: SyncMaster) {
        self.syncMaster = syncMaster
        sync()
    }

    private func sync() {
        syncMaster?.getState(
            onSuccess: { [weak self] res in
                self?.jsonState = res as? String
                self?.changeListeners.forEach { $0.onSyncedBlocStateChange() }
            },
            onError: {
                fatalError("Error")
            }
        )
    }

    func addEvent(event: String) {
        syncMaster?.addEvent(event: event)
    }
}
